lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1602,7 +1612,13 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
 
     def _change_datatype(self, item: HistoryItemModel, params: ChangeDatatypeOperationParams):
         if isinstance(item, HistoryDatasetAssociation):
-            item.change_datatype(params.datatype)
+            self.hda_manager.ensure_can_change_datatype(item)
+            self.hda_manager.ensure_can_set_metadata(item)
+            item.dataset.state = item.dataset.states.SETTING_METADATA
+            # Chain tasks using immutable signature to avoid messing with dependency injection
+            change_datatype_task = change_datatype.si(dataset_id=item.id, datatype=params.datatype)
+            set_metadata_task = set_metadata.si(dataset_id=item.id)
+            (change_datatype_task | set_metadata_task).delay()
```

Comment:
There's a cost to having separate tasks, I think `change_datatype` should just do it all in one task.